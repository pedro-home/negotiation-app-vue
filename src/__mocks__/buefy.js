import Buefy from 'buefy'

export default {
  ...Buefy,
  install: function(Vue, config) {
    Buefy.install(Vue, {
      ...config,
      defaultContainerElement: 'body',
    })
  },
}
