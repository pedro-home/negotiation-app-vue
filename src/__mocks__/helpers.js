import { mount } from '@vue/test-utils'

export const resetDOM = () => {
  document.head.innerHTML = ''
  document.body.innerHTML = ''
}

export const render = (Component, props) => {
  resetDOM()

  return mount(Component, {
    propsData: props,
    attachToDocument: true,
  })
}
