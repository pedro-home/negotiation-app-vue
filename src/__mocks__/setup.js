import Vue from 'vue'
import Buefy from 'buefy'

// Makes sure that Jest uses buefy
Vue.use(Buefy)
