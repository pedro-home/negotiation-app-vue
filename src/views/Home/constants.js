/**
 * API url for fetching temperature in London, UK
 */
export const WEATHER_API =
  'https://api.openweathermap.org/data/2.5/weather?q=London&units=metric&appid=fc8047f2d08ecbf4001369f8ee3fe543'
