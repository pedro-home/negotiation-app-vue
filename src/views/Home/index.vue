<template>
  <b-tabs type="is-toggle" expanded>
    <b-tab-item label="Employer-Tab">
      <Offer
        :text="textEmployer"
        :is-active="valueEmployer === null"
        :on-submit="saveEmployerOffer"
      />
    </b-tab-item>
    <b-tab-item label="Employee-Tab">
      <Offer
        :text="textEmployee"
        :is-active="valueEmployee === null"
        :on-submit="saveEmployeeOffer"
      />
    </b-tab-item>
  </b-tabs>
</template>

<script>
import Offer from '../../components/Offer'
import { WEATHER_API } from './constants'

export default {
  data: () => {
    return {
      textEmployer: 'Enter maximum salary',
      textEmployee: 'Enter minimum salary',
      valueEmployer: null,
      valueEmployee: null,
    }
  },
  methods: {
    saveEmployerOffer(value) {
      this.valueEmployer = parseFloat(value)
      this.validateResult()
    },
    saveEmployeeOffer(value) {
      this.valueEmployee = parseFloat(value)
      this.validateResult()
    },
    validateResult() {
      if (!this.valueEmployer || !this.valueEmployee) {
        // Still waiting for both to answer
        return
      }

      // Fetch temperature from API
      this.fetchTemperature().then((temperature) => {
        const message = `Maximum offer was: ${this.valueEmployer}<br>Minimum expected salary was: ${this.valueEmployee}<br>Current temperature: ${temperature} Â°C`

        if (this.valueEmployee <= this.valueEmployer) {
          // Show success dialog
          this.$buefy.dialog.alert({ title: 'Success!', message })
          return
        }

        // Show failure dialog
        this.$buefy.dialog.alert({ title: 'Failure!', message })
      })
    },
    async fetchTemperature() {
      let temperature
      try {
        temperature = await fetch(WEATHER_API)
          .then((resp) => {
            // Validate response
            if (!resp.ok) {
              return Promise.reject(new Error(resp.statusText))
            }

            return resp.json()
          })
          .then(({ main: { temp } }) => temp)
      } catch (e) {
        temperature = 'Error'
      }

      return temperature
    },
  },
  components: {
    Offer,
  },
}
</script>
