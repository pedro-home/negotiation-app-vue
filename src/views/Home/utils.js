import { fetchData } from '@/utils'
import { WEATHER_API } from './constants'

/**
 * Fetch temperature value from London, UK
 */
export const fetchTemperature = async () => {
  const temperature = await fetchData(WEATHER_API)
    .then(({ main: { temp } }) => temp)
    .catch(() => 'Error')

  return temperature
}
