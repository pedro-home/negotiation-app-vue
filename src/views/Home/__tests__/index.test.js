import Vue from 'vue'
import { render } from '@/__mocks__/helpers'
import Home from '../index.vue'

describe('Given <Home />', () => {
  let wrapper
  let props = {}
  beforeEach(() => {
    const defaultProps = {}
    wrapper = render(Home, { ...defaultProps, ...props })
  })

  describe('When initializes', () => {
    describe('And uses default props', () => {
      it('Then should match snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
      })

      it('And should have the employer offer field', () => {
        expect(wrapper.contains('[data-test="employer"]')).toBe(true)
      })

      it('And should have the employee offer field', () => {
        expect(wrapper.contains('[data-test="employee"]')).toBe(true)
      })
    })

    describe("And interacts with employer's tab", () => {
      let tab
      beforeEach(() => {
        tab = wrapper.find('[data-test="employer"]')
      })

      describe('And submits correct value', () => {
        beforeEach(async () => {
          tab.find('input').setValue(10)
          await Vue.nextTick()

          tab.find('button').trigger('click')
          await Vue.nextTick()
        })

        it('Then should match snapshot', () => {
          expect(wrapper.element).toMatchSnapshot()
        })

        it('And should remove content', () => {
          expect(wrapper.contains('[data-test="employer"]')).toBeFalsy()
        })

        describe("And interacts with employee's tab", () => {
          let tab
          beforeEach(async () => {
            tab = wrapper.find('[data-test="employee"]')
            wrapper.find('.tabs li:not(.is-active) > a').trigger('click')
            Vue.nextTick()
          })

          describe('And submits correct value', () => {
            beforeEach(async () => {
              tab.find('input').setValue(12)
              await Vue.nextTick()

              tab.find('button').trigger('click')
              await Vue.nextTick()
            })

            it('Then should match snapshot', () => {
              expect(wrapper.element).toMatchSnapshot()
            })

            it('And should remove content', () => {
              expect(wrapper.contains('[data-test="employee"]')).toBeFalsy()
            })

            it('And should show a dialog', () => {
              expect(document.body.querySelector('.dialog')).toBeTruthy()
            })
          })
        })
      })
    })
  })
})
