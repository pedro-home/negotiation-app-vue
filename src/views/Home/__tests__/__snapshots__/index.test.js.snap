// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given <Home /> When initializes And interacts with employer's tab And submits correct value And interacts with employee's tab And submits correct value Then should match snapshot 1`] = `
<div
  class="b-tabs is-fullwidth"
>
  <nav
    class="tabs is-toggle is-fullwidth"
  >
    <ul>
      <li
        class=""
      >
        <a>
          <!---->
           
          <span>
            Employer-Tab
          </span>
        </a>
      </li>
      <li
        class="is-active"
      >
        <a>
          <!---->
           
          <span>
            Employee-Tab
          </span>
        </a>
      </li>
    </ul>
  </nav>
   
  <section
    class="tab-content is-transitioning"
  >
    <div
      class="tab-item slide-prev-leave slide-prev-leave-active"
      style=""
    />
     
    <div
      class="tab-item slide-prev-enter slide-prev-enter-active"
      style=""
    />
  </section>
</div>
`;

exports[`Given <Home /> When initializes And interacts with employer's tab And submits correct value Then should match snapshot 1`] = `
<div
  class="b-tabs is-fullwidth"
>
  <nav
    class="tabs is-toggle is-fullwidth"
  >
    <ul>
      <li
        class="is-active"
      >
        <a>
          <!---->
           
          <span>
            Employer-Tab
          </span>
        </a>
      </li>
      <li
        class=""
      >
        <a>
          <!---->
           
          <span>
            Employee-Tab
          </span>
        </a>
      </li>
    </ul>
  </nav>
   
  <section
    class="tab-content is-transitioning"
  >
    <div
      class="tab-item v-enter v-enter-active"
      style=""
    />
     
    <div
      class="tab-item"
      style="display: none;"
    >
      <div
        data-test="employee"
      >
        <div
          class="field"
        >
          <!---->
           
          <div
            class="control is-clearfix"
          >
            <input
              autocomplete="on"
              class="input"
              placeholder="Enter minimum salary"
              type="text"
            />
             
            <!---->
             
            <!---->
             
            <!---->
          </div>
           
          <!---->
        </div>
         
        <button
          class="button"
          disabled="disabled"
          type="button"
        >
          <!---->
           
          <span>
            Submit
          </span>
           
          <!---->
        </button>
      </div>
    </div>
  </section>
</div>
`;

exports[`Given <Home /> When initializes And uses default props Then should match snapshot 1`] = `
<div
  class="b-tabs is-fullwidth"
>
  <nav
    class="tabs is-toggle is-fullwidth"
  >
    <ul>
      <li
        class="is-active"
      >
        <a>
          <!---->
           
          <span>
            Employer-Tab
          </span>
        </a>
      </li>
      <li
        class=""
      >
        <a>
          <!---->
           
          <span>
            Employee-Tab
          </span>
        </a>
      </li>
    </ul>
  </nav>
   
  <section
    class="tab-content is-transitioning"
  >
    <div
      class="tab-item v-enter v-enter-active"
      style=""
    >
      <div
        data-test="employer"
      >
        <div
          class="field"
        >
          <!---->
           
          <div
            class="control is-clearfix"
          >
            <input
              autocomplete="on"
              class="input"
              placeholder="Enter maximum salary"
              type="text"
            />
             
            <!---->
             
            <!---->
             
            <!---->
          </div>
           
          <!---->
        </div>
         
        <button
          class="button"
          disabled="disabled"
          type="button"
        >
          <!---->
           
          <span>
            Submit
          </span>
           
          <!---->
        </button>
      </div>
    </div>
     
    <div
      class="tab-item"
      style="display: none;"
    >
      <div
        data-test="employee"
      >
        <div
          class="field"
        >
          <!---->
           
          <div
            class="control is-clearfix"
          >
            <input
              autocomplete="on"
              class="input"
              placeholder="Enter minimum salary"
              type="text"
            />
             
            <!---->
             
            <!---->
             
            <!---->
          </div>
           
          <!---->
        </div>
         
        <button
          class="button"
          disabled="disabled"
          type="button"
        >
          <!---->
           
          <span>
            Submit
          </span>
           
          <!---->
        </button>
      </div>
    </div>
  </section>
</div>
`;
